funsor/sum_product.py

```diff
@@ -45,6 +45,67 @@ def _partition(terms, sum_vars):
     return components
 
 
+def _unroll_plate(factors, plate, var_to_ordinal):
+    # replicated variables
+    plate_vars = set()
+    for var, ordinal in var_to_ordinal.items():
+        if plate in ordinal:
+            plate_vars.add(var)
+
+    # unroll factors
+    unrolled_factors = []
+    for factor in factors:
+        if plate in factor.inputs:
+            size = factor.inputs[plate].size
+            f_vars = plate_vars.intersection(factor.inputs)
+            unrolled_factors.extend([factor(
+                    **{plate: i},
+                    **{var: "{}_{}{}".format(var, plate, i) for var in f_vars}
+                ) for i in range(size)])
+        else:
+            unrolled_factors.append(factor)
+
+    # unroll variables
+    for var in plate_vars:
+        ordinal = var_to_ordinal.pop(var)
+        new_ordinal = ordinal.difference(plate)
+        var_to_ordinal.update({"{}_{}{}".format(var, plate, i): new_ordinal
+                               for i in range(size)})
+
+    return unrolled_factors, var_to_ordinal
+
+
+def partial_unroll(factors, eliminate=frozenset(), plates=frozenset()):
+    """
+    Performs partial unrolling of plated factor graphs to standard factor graphs.
+
+    :return: a list of partially unrolled Funsors,
+        a frozenset of partially unrolled variable names,
+        and a frozenset of remaining plates.
+    """
+    assert isinstance(factors, (tuple, list))
+    assert all(isinstance(f, Funsor) for f in factors)
+    assert isinstance(eliminate, frozenset)
+    assert isinstance(plates, frozenset)
+    sum_vars = eliminate - plates
+    unrolled_plates = eliminate & plates
+
+    var_to_ordinal = {}
+    for f in factors:
+        ordinal = plates.intersection(f.inputs)
+        for var in sum_vars.intersection(f.inputs):
```

Comment:
Sorry - to clarify, I think you should still be unrolling all plates and variables, but you should only *return* an updated version of `eliminate`, so that the output of `partial_unroll` can be passed directly to `partial_sum_product` with an empty `plates` set.  As of this comment, you're no longer unrolling all variables, although you are returning `eliminate`.